### PR TITLE
`no-unknown-modifiers` rule

### PR DESCRIPTION
**Pull request by jfmengels**
Adds `no-unknown-modifiers` rule

Detects unknown `test` modifiers
```js
test.onlu(...)
test.ski(...)
test.Before(...)
test.beforeeach(...)
```

I have omitted 'default' and 'test' from the modifier list, as I think those are simply import noise. Let me know what you think.

### CODE REVIEW

**Comment by jamestalmage** [comment]
Hmm. I guess I don't get the value. These types of mistakes are going to surface at runtime. Why do all this work to detect a problem that will be detected anyways?

**Comment by jfmengels** [comment]
This rule is only useful when you have a linter in your editor, and it's not the only rule that has been written for that purpose (like: `no-invalid-end`).
I agree that #54 is pretty useless as it pretty much never happens, but catching this kind of common error (typos...) very early is definitely useful IMO. Outside of an editor, I agree that it has no value.

**Comment by sindresorhus** [comment]
I think it would be better to just hardcode the list of modifiers. It's not going to get updated very often. Hopefully never after 1.0.0. And we control both so we can easily update. Requiring `ava` comes with unnecessary overhead.

**Comment by sindresorhus** [comment]
I first thought what James said, but after thinking about it, it's handy for catching early errors.

**Comment by jamestalmage** [comment]
:+1:

**Comment by jfmengels** [comment]
Ok, I will hardcode the modifiers then. I'll probably make a similar rule for assertions later on.

**Comment by jfmengels** [comment]
Updated the PR and its description. It's not a WIP anymore.

File: rules/no-unknown-modifiers.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	after: true,
+	afterEach: true,
+	before: true,
+	beforeEach: true,
+	cb: true,
+	only: true,
+	serial: true,
+	skip: true,
+	todo: true
+};
```
Comment (sindresorhus) [inline rules/no-unknown-modifiers.js]: Why not just an array? You won't gain any noticeable speed improvement for such a small list.

File: rules/no-unknown-modifiers.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+function getTestModifiers(node) {
+	if (node.type === 'CallExpression') {
+		return getTestModifiers(node.callee);
+	}
```
Comment (sindresorhus) [inline rules/no-unknown-modifiers.js]: empty line below all blocks

File: rules/no-unknown-modifiers.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	if (node.type === 'MemberExpression') {
+		return getTestModifiers(node.object).concat(node.property.name);
+	}
```
Comment (sindresorhus) [inline rules/no-unknown-modifiers.js]: empty line

File: rules/no-unknown-modifiers.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+function unknownModifiers(node) {
+	return getTestModifiers(node)
+		.filter(function (modifier) {
+			return !modifiers[modifier];
+		});
+}
+
+module.exports = function (context) {
+	var ava = createAvaRule();
```
Comment (sindresorhus) [inline rules/no-unknown-modifiers.js]: empty line below here

File: rules/no-unknown-modifiers.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	return ava.merge({
+		CallExpression: function (node) {
+			if (!ava.isTestFile || ava.currentTestNode !== node) {
+				return;
+			}
```
Comment (sindresorhus) [inline rules/no-unknown-modifiers.js]: empy line

File: rules/no-unknown-modifiers.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+			var unknown = unknownModifiers(node);
```
Comment (sindresorhus) [inline rules/no-unknown-modifiers.js]: empty line

File: rules/no-unknown-modifiers.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+			if (unknown.length !== 0) {
+				context.report(node, 'Unknown modifier "' + unknown[0] + '"');
```
Comment (sindresorhus) [inline rules/no-unknown-modifiers.js]: Use 

```
`
```

instead of `"`

File: test/no-unknown-modifiers.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+			`${header} test.skip(t => {});`,
+			`${header} test.todo(t => {});`,
+			// shouldn't be triggered since it's not a test file
+			`test.foo(t => {});`
+		],
+		invalid: [
+			{
+				code: `${header} test.foo(t => {});`,
+				errors: [
+          {...ruleError, message: 'Unknown modifier "foo"'}
```
Comment (sindresorhus) [inline test/no-unknown-modifiers.js]: inconsidtent indent

File: rules/no-unknown-modifiers.js (hunk 1 above)
Comment (jfmengels) [inline rules/no-unknown-modifiers.js]: Was not a performance related decision. I found `modifiers[x]` was just a pretty way to know whether x is in there. But yeah, it's more verbose in the list declaration bit, and a tad more error-prone (`{ ..., value: false, ...`). Changed.

File: rules/no-unknown-modifiers.js (hunk 2 above)
Comment (jfmengels) [inline rules/no-unknown-modifiers.js]: Sure thing, done ;)
I'd say make it a xo rule, but it's not a rule that I like much, and I'm starting to use xo more and more, so I won't say anything ;)

**Comment by sindresorhus** [comment]
LGTM

**Comment by jfmengels** [comment]
Updated

**Comment by sindresorhus** [comment]
Thanks @jfmengels :)

Internet dropped so couldn't post a gif yesterday.

leo-excited

**Pull request closed without merging**